modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/defragmentation/IgniteDefragmentationImpl.java

```diff
@@ -0,0 +1,223 @@
+            }
+            catch (Exception e) {
+                throw new IgniteCheckedException("Exception occurred: " + e.getMessage(), e);
+            }
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public DefragmentationStatus status() throws IgniteCheckedException {
+        final MaintenanceRegistry maintenanceRegistry = ctx.maintenanceRegistry();
+
+        if (!maintenanceRegistry.isMaintenanceMode())
+            throw new IgniteCheckedException("Node is not in maintenance mode.");
+
+        IgniteCacheDatabaseSharedManager dbMgr = ctx.cache().context().database();
+
+        assert dbMgr instanceof GridCacheDatabaseSharedManager;
+
+        CachePartitionDefragmentationManager defrgMgr = ((GridCacheDatabaseSharedManager)dbMgr)
+                .defragmentationManager();
+
+        if (defrgMgr == null)
+            throw new IgniteCheckedException("There's no active defragmentation process on the node.");
+
+        final Status status = defrgMgr.status();
+
+        final long startTs = status.getStartTs();
+        final long finishTs = status.getFinishTs();
+        final long elapsedTime = finishTs != 0 ? finishTs - startTs : System.currentTimeMillis() - startTs;
+
+        Map<String, CompletedDefragmentationInfo> completedCaches = new HashMap<>();
+        Map<String, InProgressDefragmentationInfo> progressCaches = new HashMap<>();
+
+        status.getFinishedGroups().forEach((context, progress) -> {
+            final String name = context.cacheOrGroupName();
+
+            final long oldSize = progress.getOldSize();
+            final long newSize = progress.getNewSize();
+            final long cgElapsedTime = progress.getFinishTs() - progress.getStartTs();
+
+            final CompletedDefragmentationInfo info = new CompletedDefragmentationInfo(cgElapsedTime, oldSize, newSize);
+            completedCaches.put(name, info);
+        });
+
+        status.getProgressGroups().forEach((context, progress) -> {
+            final String name = context.cacheOrGroupName();
+
+            final long cgElapsedTime = System.currentTimeMillis() - progress.getStartTs();
+            final int partsTotal = progress.getPartsTotal();
+            final int partsCompleted = progress.getPartsCompleted();
+
+            final InProgressDefragmentationInfo info = new InProgressDefragmentationInfo(cgElapsedTime, partsCompleted, partsTotal);
+            progressCaches.put(name, info);
+        });
+
+        return new DefragmentationStatus(
+            completedCaches,
+            progressCaches,
+            status.getScheduledGroups(),
+            status.getSkippedGroups(),
+            status.getTotalPartitionCount(),
+            status.getDefragmentedPartitionCount(),
+            startTs,
+            elapsedTime
+        );
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean inProgress() {
+        final Status status = getStatus();
+
+        return status != null && status.getFinishTs() == 0;
```

Comment:
So, if status is null then inProgress is false.
1) is that correct?
2) how is it possible?

Is this the case when we're not in maintanance mode or there's no defragmentation scheduled?

Reply:
Yes, status is null if the node is not in the maintenance mode/no defragmentation scheduled